Tests/AsyncHTTPClientTests/AsyncTestHelpers.swift

```diff
@@ -0,0 +1,189 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AsyncHTTPClient open source project
+//
+// Copyright (c) 2021 Apple Inc. and the AsyncHTTPClient project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AsyncHTTPClient project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+#if swift(>=5.5) && canImport(_Concurrency)
+import NIOConcurrencyHelpers
+import NIOCore
+
+@available(macOS 12.0, iOS 15.0, watchOS 8.0, tvOS 15.0, *)
+class AsyncSequenceWriter: AsyncSequence {
+    typealias AsyncIterator = Iterator
+    typealias Element = ByteBuffer
+
+    struct Iterator: AsyncIteratorProtocol {
+        typealias Element = ByteBuffer
+
+        private let writer: AsyncSequenceWriter
+
+        init(_ writer: AsyncSequenceWriter) {
+            self.writer = writer
+        }
+
+        mutating func next() async throws -> ByteBuffer? {
+            try await self.writer.next()
+        }
+    }
+
+    func makeAsyncIterator() -> Iterator {
+        return Iterator(self)
+    }
+
+    private enum State {
+        case buffering(CircularBuffer<ByteBuffer?>, CheckedContinuation<Void, Error>?)
+        case finished
+        case waiting(CheckedContinuation<ByteBuffer?, Error>)
+        case failed(Error)
+    }
+
+    private var _state = State.buffering(.init(), nil)
+    private let lock = Lock()
+
+    public var hasDemand: Bool {
+        self.lock.withLock {
+            switch self._state {
+            case .failed, .finished, .buffering:
+                return false
+            case .waiting:
+                return true
+            }
+        }
+    }
+
+    public func demand() async throws {
```

Comment:
What is this method used for? Does it wait until someone calls next but the buffer is empty?

Reply:
Added a comment.